blockchain/blockchain.go

```diff
@@ -608,82 +667,225 @@ func (b *Blockchain) WriteHeadersWithBodies(headers []*types.Header) error {
 	return nil
 }
 
-// WriteBlock writes a single block
-func (b *Blockchain) WriteBlock(block *types.Block) error {
-	// Check the param
+// VerifyProposedBlock does the minimal block verification without consulting the
+// consensus layer. Should only be used if consensus checks are done
+// outside the method call
+func (b *Blockchain) VerifyProposedBlock(block *types.Block) error {
+	// Do just the initial block verification
+	return b.verifyBlock(block)
+}
+
+// VerifySealedBlock verifies that the block is valid by performing a series of checks.
+// It is assumed that the block status is sealed (committed)
+func (b *Blockchain) VerifySealedBlock(block *types.Block) error {
+	// Do the initial block verification
+	if err := b.verifyBlock(block); err != nil {
+		return err
+	}
+
+	// Make sure the consensus layer verifies this block header
+	if err := b.consensus.VerifyHeader(block.Header); err != nil {
+		return fmt.Errorf("failed to verify the header: %w", err)
+	}
+
+	return nil
+}
```

Comment:
Yes - I completely agree with you.

It's part of the technical debt that is accumulating, that the `blockchain` module needs to double check with the consensus module on the validity of a block 🤦‍♂️ 

Currently, there is no way to circumvent this. If you open up this `VerifyHeader` method, you'll notice that it checks for the committed seals - something you don't have when verifying the block from the point of consensus `AcceptState` - that's why block verification is split up into these 2 methods (`VerifySealedBlock` and `VerifyProposedBlock`)

Reply:
At least, the naming of `VerifySealedBlock` and `VerifyProposedBlock` depends on IBFT consensus completely. We need to think about renaming these method.

Reply:
What do you propose we name these?

Reply:
It's much cleaner to call 2 methods in consensus rather than calling specific method of Blockchain

Reply:
I'm not sure I understand what you mean, we already call 2 methods in consensus:
1. During `AcceptState` - > `VerifyProposedBlock`
2. During `CommitState` -> `WriteBlock`

Reply:
The problem is that Blockchain has 2 methods to verify block. It completely depends on consensus. The method names breaks blockchain module. The name must explain what this method does. If you can't, you need to remove `VerifyProposedBlock` and call all steps in `VerifyProposedBlock` from consensus directly.

Reply:
I don't think adding these 2 methods break the blockchain module - it was broken the moment we introduced and tied blockchain functionality to the consensus layer:
* [PreStateCommits](https://github.com/0xPolygon/polygon-edge/blob/ddc34803e0ce1f28fc1999023cc02a9248ddd6e4/blockchain/blockchain.go#L844-L846)
* [Snapshot updating](https://github.com/0xPolygon/polygon-edge/blob/ddc34803e0ce1f28fc1999023cc02a9248ddd6e4/blockchain/blockchain.go#L898-L900)
* [Header verification](https://github.com/0xPolygon/polygon-edge/blob/ddc34803e0ce1f28fc1999023cc02a9248ddd6e4/blockchain/blockchain.go#L672-L674)
* [Block creator fetches](https://github.com/0xPolygon/polygon-edge/blob/ddc34803e0ce1f28fc1999023cc02a9248ddd6e4/blockchain/blockchain.go#L834-L837)

Let's not pretend that all of these methods are not IBFT specific and endanger the independence of the blockchain module.

We're discussing if having 2 methods breaks some kind of blockchain independence, when it was broken to begin with, before we added this separation of concern.

The consensus module cannot do the block verification internally because it would require a much bigger separation of concern (primarily of the executor and blockchain module), and additional exporting of blockchain specific methods for this verification to actually happen within consensus.

The best we can do, as of this task and the scope of this task, is to rename these methods to something more meaningful. What names do you suggest?

Reply:
I don't think the examples you're raising are different from this. 

Sure they have been added because consensus needed, but they are just delegation and it means giving some limited ability to inject call. I don't think this is same.

btw, I agree with your opinion. All I can do right now is to rename method.

Reply:
That sounds reasonable to me 👍 

Renamed:

[dc73808](https://github.com/0xPolygon/polygon-edge/pull/544/commits/dc73808a50978b603b72e409e9c4b1a19dd4cfb2)